Core/include/Acts/AmbiguityResolution/AmbiguityResolutionML.hpp

```diff
@@ -0,0 +1,126 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Units.hpp"
+#include "Acts/EventData/TrackContainer.hpp"
+#include "Acts/Utilities/Delegate.hpp"
+#include "Acts/Utilities/Logger.hpp"
+
+#include <cstddef>
+#include <map>
+#include <memory>
+#include <string>
+#include <tuple>
+#include <vector>
+
+namespace Acts {
+
+/// Generic implementation of the machine learning ambiguity resolution
+/// Contains method for data preparations
+template <typename AmbiguityNetwork>
```

Comment:
This seems very generic to me. It would be great to either have a virtual interface or a C++20 concept that further constrains the required type, also maybe some more documentation.
As a third alternative, also 2 `Acts::Delegate` could be a good interface, assuming the algorithm uses two independent neural networks.

Reply:
Good idea, I haven't had the time to fully study the c++20 concept but it is probably a great opportunity !

Reply:
This now uses a AmbiguityNetworkConcept